src/Operation/Aggregate.php

```diff
@@ -138,18 +138,9 @@ class Aggregate implements Executable, Explainable
      */
     public function __construct(string $databaseName, ?string $collectionName, array $pipeline, array $options = [])
     {
-        $expectedIndex = 0;
-
-        foreach ($pipeline as $i => $operation) {
-            if ($i !== $expectedIndex) {
-                throw new InvalidArgumentException(sprintf('$pipeline is not a list (unexpected index: "%s")', $i));
-            }
-
-            if (! is_array($operation) && ! is_object($operation)) {
-                throw InvalidArgumentException::invalidType(sprintf('$pipeline[%d]', $i), $operation, 'array or object');
-            }
-
-            $expectedIndex += 1;
+        // Empty pipelines are allowed
+        if ($pipeline !== [] && ! is_pipeline($pipeline)) {
```

Comment:
`is_pipeline` returns `false` for an empty list. There is a test that assert an empty pipeline is valid `AggregateFunctionalTest::testEmptyPipelineReturnsAllDocuments`.

I wonder if empty list is allowed elsewhere and if the constraint should be relaxed in `is_pipeline` to be explicitly stricter when necessary.

Reply:
I'm on the fence about this. In general I'd say that `is_pipeline` should return `true` for anything that is a valid aggregation pipeline, which includes an empty array. In those cases where we want to forbid using an empty pipeline, we should explicitly guard against a non-empty array. Thoughts @jmikola?

Reply:
Note that `is_pipeline()` was only created for validation of `update` and `findAndModify` command options (not `aggregate`). The original prohibition for empty arrays dates back to https://github.com/mongodb/mongo-php-library/commit/2628c098e3bfc015d7e2d59d1437cf1fed046e12#diff-e093a0fd8f57549d81ad2871e6ca3872ecaef99d0525118e26c0501bdd0c1d92R241 from [PHPLIB-418](https://jira.mongodb.org/browse/PHPLIB-418).

I don't recall why that was done. I recently confirmed in [DRIVERS-2638](https://jira.mongodb.org/browse/DRIVERS-2638) that empty pipelines are accepted for updates.